WIP: adding csiconfig api and controller

peterochodo opened the pull request:
### What this PR does / why we need it
- so that addon manager can watch and process CSIConfig objects --> which are necessary for configuring vsphere-csi and vsphere-pv-csi addons

### Describe testing done for PR

- created a Cluster object, created a CSIConfig object with the same name as Cluster, verified that expected secret was created with the expected data
- deleted Cluster object and verified that CSIConfig object and corresponding secret were also deleted
- created CSIConfig object without corresponding Cluster object and verified that it was ignored

Fixes #1579 

### Fix
- https://github.com/vmware-tanzu/tanzu-framework/issues/1579

### Release note
- enables configuration of vsphere-pv-csi and vsphere-csi addons via CSIConfig api

Commit:
adding csiconfig api and controller

stmcginnis commented on addons/controllers/csi/csiconfig_constants.go:
```diff
@@ -0,0 +1,23 @@
+/*
+
```
Needs VMware copyright header.

peterochodo replied:
done

stmcginnis commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+/*
+Copyright 2022.
```
Need full header.

peterochodo replied:
updated

stmcginnis commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+
```
Extra line break should be removed.

peterochodo replied:
updated

stmcginnis reviewed (commented):
Just pointing out a couple nits inline. Also looks like this branch was maybe created from a stale `main` branch as their is already a merge conflict.

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+func (r *CSIConfigReconciler) reconcileCSIConfig(ctx context.Context,
+	csiCfg *csiv1alpha1.CSIConfig) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
```
Why not just using the logger from controller runtime as a field inside ```CSIConfigReconciler```?

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch CSIConfig")
+			return
```
Where do you set ```retErr```? Shouldn't you set ```retErr``` here to ```err```?

blc1996 replied:
The usage of `retErr` is a bit implicit here, not sure if I follow correctly. Recommend some refactor.

maralavi replied:
I'd expect ```retErr = err``` on line 112 for the ```retErr``` explicitly named return value to have any meaning

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch CSIConfig")
+			return
+		}
```
Could you please log a successful patch here?

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+
+	if result, err = r.reconcileCSIConfigNormal(ctx, csiCfg); err != nil {
+		logger.Error(err, "Error reconciling CSIConfig")
+		return result, err
```
On error you should return empty object ```ctrl.Result{}```

peterochodo replied:
reconcileCSIConfigNormal makes sure that result is the appropriate value. It's a delegate function, so we trust it will do the right thing & thus return its results.

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+		logger.Error(err, "Error reconciling CSIConfig")
+		return result, err
+	}
+
```
Could you please log successful reconciliation here?

peterochodo replied:
- the approach being used is 'log only if their is a problem'. Otherwise you will have very noisy logs in a production system.

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+	logger := log.FromContext(ctx)
+
+	// add owner reference to CSIConfig if not already added
+	cluster, err := r.getOwnerCluster(ctx, csiCfg)
```
Why not passing ```cluster``` from main Reconcile() and not getting it again here?

peterochodo replied:
- it was effort to keep the code to look 'clean': Note that 'cluster' is used again 2 delegate functions down the stack (in reconcileVSphereCSIConfigNormal). getOwnerCluster function seems relatively cheap to invoke.

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+func (r *CSIConfigReconciler) reconcileCSIConfigNormal(ctx context.Context,
+	csiCfg *csiv1alpha1.CSIConfig) (result ctrl.Result, retErr error) {
```
Why do you need ```retErr``` here, do you set or use it somewhere?

maralavi replied:
Same question for variable ```result```

peterochodo replied:
updated

maralavi commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+
+	if err != nil {
+		logger.Error(err, "Error creating or patching CSIConfig data values secret")
+		return result, err
```
What is the value of ```result```? I don't see it being set anywhere. Is that always ```ctrl.Result{}```?

peterochodo replied:
updated

saji-pivotal committed:
Added validation to set Azure SSH Public Key (#1841)

Signed-off-by: Sudarshan <asudarshan@vmware.com>

blc1996 commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+	addonName := ""
+	switch csiCfg.Spec.VSphereCSI.Mode {
+	case CSINonParavirtualMode:
+		addonName = "vsphere-csi"
```
`vsphere-csi` and `vsphere-pv-csi` can be stored as constants

peterochodo replied:
updated

blc1996 commented on addons/controllers/csi/csiconfig_datavalues.go:
```diff
@@ -0,0 +1,53 @@
+	PublicNetwork         string  `json:"publicNetwork"`
+	Username              string  `json:"username"`
+	Password              string  `json:"password"`
+	Region                *string `json:"region,omitempty"`
```
Non-blocking: Using pointer types will introduce more trouble when accessing them (nilness check). Using value type still work with `omitempty`, as for string, "" will be omitted when being marshalled.

peterochodo replied:
updated

blc1996 commented on addons/controllers/csi/csiconfig_utils.go:
```diff
@@ -0,0 +1,125 @@
+)
+
+// mapCSIConfigToDataValues maps CSIConfig CR to data values
+func (r *CSIConfigReconciler) mapCSIConfigToDataValues(ctx context.Context, csiConfig *csiv1alpha1.CSIConfig) (*DataValues, error) {
```
A overall question for all the map functions: Have we visited all the values and decided what are derived fields? Looks like all values come directly from config CR.

peterochodo replied:
updated

blc1996 commented on addons/main.go:
```diff
@@ -233,6 +236,13 @@ func enableClusterBootstrapAndConfigControllers(ctx context.Context, mgr ctrl.Ma
 	}
+	if err := (&csicontroller.CSIConfigReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
```
The log object should be passed in and used.

peterochodo replied:
it's being extracted from the context

blc1996 reviewed (commented):
Blocking: Need to add CSI test cases to the env test.

vuil committed:
Ensure no_proxy entries are sorted

Fixes: #1877

Signed-off-by: Vui Lam <vui@vmware.com>

Commit:
Adds function to continuously manage webhook TLS certificates (#1803)

- New self signed server certificates, key, and ca cert are created
  and placed in a k8s opaque secret.
- The server certificate and key are written to the specified path
- Webhook Mutating and Validating configurations are augmented with new CA bundle
- The new certificates are monitored for validity and rotated when appropriate.
- The new certificates are written to a namespaced secret
- Maximum allowed value for rotation time is one week
- Minimum allowed value for rotation is > 0

Commit:
Disable TKR API conversion webhook for now

Signed-off-by: Ivan Mikushin <imikushin@vmware.com>

Commit:
Increase timeout running make test

Signed-off-by: Ivan Mikushin <imikushin@vmware.com>

HanFa committed:
Create or update ProviderServiceAccount for CPI para-virtual mode. (#1868)

ggpaue committed:
fix addons-manager-v2 package sha (#1870)

Signed-off-by: Guanpeng Gao <gguanpeng@vmware.com>

imikushin committed:
TKR Resolver: Cluster webhook (#1835)

TKR Resolver: Cluster webhook

Also:
- Tweak data.Query to be able to skip resolution.
- Make data.Query structure reflect that of Cluster topology more directly: use a slice of machine deployment query parts vs a map.
- Set TKR_KUBERNETES_SPEC cluster variable: if the TKR_KUBERNETES_SPEC variable is defined in the ClusterClass,
set it to the spec.kubernetes value of the resolved TKR.
- Fix a panic in the TKR Resolver implementation when non-existent OSImage is listed by a TKR.

Signed-off-by: Ivan Mikushin <imikushin@vmware.com>

imikushin committed:
Add kube-vip field to TKR.spec.kubernetes (#1874)

Add kube-vip field to TKR.spec.kubernetes

Signed-off-by: Ivan Mikushin <imikushin@vmware.com>

Commit:
adding csiconfig controller integration tests

vijaykatam commented on apis/csi/v1alpha1/csiconfig_types.go:
```diff
@@ -0,0 +1,155 @@
+
+// CSIConfigSpec defines the desired state of CSIConfig
+// CSIConfigSpec defines the desired state of CSIConfig
+type CSIConfigSpec struct {
```
We need to prefix API types and controlelr with name with Vsphere similar to vspherecpi

peterochodo replied:
updated

sabbey37 committed:
Initial pinniped config controller commit (#1884)

Commit:
renaming CSIConfig to VSphereCSIConfig

Commit:
renaming CSIConfig to VSphereCSIConfig

Commit:
collection of code-review feedback changes

peterochodo commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
```
updated

peterochodo commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *CSIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("CSIConfig", req.NamespacedName)
```
ok

maralavi replied:
@peterochodo FYI, the decision for refactoring postponed for the future and for now the decision is to keep all addons controllers' code consistent in this respect.

peterochodo commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+	csiConfig := &csiv1alpha1.CSIConfig{}
+	if err := r.Get(ctx, req.NamespacedName, csiConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("CSIConfig resource not found", "namespacedName", req.NamespacedName)
```
updated

peterochodo commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+		return ctrl.Result{}, err // no need to requeue if cluster is not found
+	}
+
+	return r.reconcileCSIConfig(ctx, csiConfig)
```
- the csiConfig object is created, populated and owned by Reconcile function. So it's ok (& expected) if a delegate function of Reconcile updates it.
- the feared error would only be a risk if multiple 'Reconcile' function instances where concurrently being invoked on the very same csiconfig object (a race condition). However, in main.go we have setup the controller manager to max-concurrent-reconciles == 1 (as seen below). So this should not be a problem.

if err := (&csicontroller.VSphereCSIConfigReconciler{
		Client: mgr.GetClient(),
		Scheme: mgr.GetScheme(),
	}).SetupWithManager(ctx, mgr, controller.Options{MaxConcurrentReconciles: 1}); err != nil {
		setupLog.Error(err, "unable to create VSphereCSIConfigController", "controller", "csi")
		os.Exit(1)
	}

maralavi replied:
Thanks for the clarification @peterochodo

peterochodo commented on addons/controllers/csi/csiconfig_controller.go:
```diff
@@ -0,0 +1,196 @@
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Unable to fetch CSIConfig resource", "namespacedName", req.NamespacedName)
```
updated

adobley committed:
Configure vSphere CPI to exclude kube-vip IP (#1480)

To prevent vSphere CPI from selecting the kube-vip IP as the node's IP,
configure it to be excluded from selection. The kube-vip IP is assigned
from the VSPHERE_CONTROL_PLANE_ENDPOINT config variable.

Recent changes to the vSphere CPI allow it to be configured to exclude
CIDRs from selection.

Co-authored-by: Aidan Obley <aobley@vmware.com>
Co-authored-by: Christian Ang <angc@vmware.com>
Co-authored-by: Tyler Schultz <tschultz@vmware.com>

benjaminapetersen committed:
Add Makefile target to generate the default pinniped addon secret (#1892)

Signed-off-by: Benjamin A. Petersen <ben@benjaminapetersen.me>

benjaminapetersen committed:
Makefile/additions/controller-gen and kustomize (#1918)

* Add controller-gen and kustomize to /hack/tools and update Makefile

Signed-off-by: Benjamin A. Petersen <ben@benjaminapetersen.me>

* Regenerate configs

Signed-off-by: Benjamin A. Petersen <ben@benjaminapetersen.me>

sabbey37 committed:
addons/pinniped/config-controller: Watch clusterbootstrap secrets (#1908)

saji-pivotal committed:
Removed slash from the custom image repo being read in as hostname (#1922)

codegold79 committed:
Plugin description wraps if its column exceeds 80 characters (#1839)

Signed-off-by: F. Gold <fgold@vmware.com>

codegold79 committed:
Not using WithVersion no longer causes error and allows any version (#1724)

Signed-off-by: F. Gold <fgold@vmware.com>

iXinqi committed:
add antrea Egress and FlowExporter params (#1869)

scothis committed:
Load all credential helpers (#1816)

Any place we load one of the k8s credential helpers, load them all.
Since we don't know what clusters users will use Tanzu with, we need to
support everything.

Signed-off-by: Scott Andrews <andrewssc@vmware.com>

Commit:
populating csi config with derived values from cluster

Commit:
re-adding changes lost after rebase

Commit:
Merge branch 'main' into adding-csiconfig-api-and-controller

peterochodo commented:
NOTE: I will be cancelling this MR and creating a new one --> while synching the forked main branch with tanzu-framework then attempting to rebase changes on it, something broke, so my changes are interspersed with synched changes.

peterochodo commented:
replaced by https://github.com/vmware-tanzu/tanzu-framework/pull/1942
